File: console/src/lib/replay.ts
// Replay 播放器逻辑：回放包解析、哈希链离线验证（WebCrypto）、时间线/帧对齐。
// 与内核 journal 同构：hash = sha256(prev + raw)，prev 链接前行 hash。

import type { JournalEntry } from "./journal";

export interface ReplayLine {
  seq: number;
  prev: string;
  hash: string;
  raw: string;
}

export interface ReplayFrame {
  ts_ms: number;
  format: string;
  data_base64: string;
}

export interface ReplayBundle {
  format_version: number;
  pid: string;
  engine: string;
  exported_at_ms: number;
  journal: ReplayLine[];
  frames: ReplayFrame[];
}

/** 链验证报告。 */
export interface ChainReport {
  ok: boolean;
  checked: number;
  /** 首个断链/篡改行的 seq（ok 时为 null）。 */
  brokenAt: number | null;
  reason: string | null;
}

/** 解析回放包 JSON（文件或 obs.replay.export 返回）。畸形输入抛 Error。 */
export function parseBundle(raw: unknown): ReplayBundle {
  const o = (typeof raw === "string" ? JSON.parse(raw) : raw) as Record<string, unknown>;
  if (!o || typeof o !== "object") throw new Error("replay bundle must be an object");
  if (o.format_version !== 1) {
    throw new Error(`unsupported replay format_version: ${String(o.format_version)}`);
  }
  if (typeof o.pid !== "string" || !Array.isArray(o.journal)) {
    throw new Error("replay bundle missing pid/journal");
  }
  const journal: ReplayLine[] = [];
  for (const l of o.journal) {
    const r = l as Record<string, unknown>;
    if (
      typeof r.seq === "number" &&
      typeof r.prev === "string" &&
      typeof r.hash === "string" &&
      typeof r.raw === "string"
    ) {
      journal.push({ seq: r.seq, prev: r.prev, hash: r.hash, raw: r.raw });
    }
  }
  const frames: ReplayFrame[] = [];
  if (Array.isArray(o.frames)) {
    for (const f of o.frames) {
      const r = f as Record<string, unknown>;
      if (typeof r.ts_ms === "number" && typeof r.data_base64 === "string") {
        frames.push({
          ts_ms: r.ts_ms,
          format: typeof r.format === "string" ? r.format : "png",
          data_base64: r.data_base64,
        });
      }
    }
  }
  return {
    format_version: 1,
    pid: o.pid,
    engine: typeof o.engine === "string" ? o.engine : "?",
    exported_at_ms: typeof o.exported_at_ms === "number" ? o.exported_at_ms : 0,
    journal,
    frames,
  };
}

async function sha256hex(text: string): Promise<string> {
  const data = new TextEncoder().encode(text);
  const digest = await crypto.subtle.digest("SHA-256", data);
  return Array.from(new Uint8Array(digest))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

/**
 * 离线重放验证链段：
 * - 首行之后每行 `prev` 必须等于前行 `hash`（链接性）
 * - 每行 `hash` 必须等于 sha256(prev + raw)（完整性）
 */
export async function verifyChain(journal: ReplayLine[]): Promise<ChainReport> {
  let prevHash: string | null = null;
  for (const line of journal) {
    if (prevHash !== null && line.prev !== prevHash) {
      return {
        ok: false,
        checked: journal.length,
        brokenAt: line.seq,
        reason: "链断裂：prev 与前行 hash 不一致",
      };
    }
    const expect = await sha256hex(line.prev + line.raw);
    if (expect !== line.hash) {
      return {
        ok: false,
        checked: journal.length,
        brokenAt: line.seq,
        reason: "hash 不匹配：行内容被修改",
      };
    }
    prevHash = line.hash;
  }
  return { ok: true, checked: journal.length, brokenAt: null, reason: null };
}

/** 时间线条目：解析 raw 为 journal entry，标记是否属于目标 pid。 */
export interface TimelineItem {
  entry: JournalEntry;
  ofPid: boolean;
}

/** 链段 → 时间线（旧→新）；raw 不可解析的行跳过。 */
export function timeline(bundle: ReplayBundle): TimelineItem[] {
  const out: TimelineItem[] = [];
  for (const line of bundle.journal) {
    try {
      const e = JSON.parse(line.raw) as JournalEntry;
      if (typeof e.seq !== "number") continue;
      out.push({ entry: e, ofPid: e.pid === bundle.pid });
    } catch {
      // 跳过畸形行（验证器会单独报告链完整性）
    }
  }
  return out;
}

/** 给定时间点，找到不晚于该时刻的最近一帧（无帧返回 null）。 */
export function frameAt(frames: ReplayFrame[], tsMs: number): ReplayFrame | null {
  let best: ReplayFrame | null = null;
  for (const f of frames) {
    if (f.ts_ms <= tsMs && (best === null || f.ts_ms > best.ts_ms)) best = f;
  }
  return best;
}

/** 帧 → data URL。 */
export function frameUrl(frame: ReplayFrame): string {
  return `data:image/${frame.format};base64,${frame.data_base64}`;
}

File: crates/scootlens-abi/src/replay.rs
//! 回放包类型（`obs.replay.export`，docs/03-abi-spec.md）。
//!
//! 回放包 = journal 哈希链**连续段**（未按 pid 过滤，可离线重放验链）加上
//! 画面帧序列（内核 FrameStore 采集）。Console Replay 播放器离线加载：
//! 逐行校验 `hash == sha256(prev + raw)` 与 `prev` 链接，再按 pid 过滤时间线。

use serde::{Deserialize, Serialize};

use crate::id::Pid;

/// 回放包格式版本（破坏性变更时递增）。
pub const REPLAY_FORMAT_VERSION: u32 = 1;

/// `obs.replay.export` 返回的回放包。
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ReplayBundle {
    /// 包格式版本（当前 1）。
    pub format_version: u32,
    /// 目标进程。
    pub pid: Pid,
    /// 导出时内核引擎标识。
    pub engine: String,
    /// 导出时间（unix 毫秒）。
    pub exported_at_ms: u64,
    /// journal 哈希链连续尾段（旧→新）。含全部主体/方法（已脱敏），
    /// 离线验证：首行之后每行 `prev` 必须等于前行 `hash`。
    pub journal: Vec<ReplayLine>,
    /// 画面帧（旧→新；与 journal 条目按 `ts_ms` 对齐）。
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub frames: Vec<ReplayFrame>,
}

/// journal 链上一行（与内核 journal.jsonl 行格式一致）。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ReplayLine {
    /// 单调序号。
    pub seq: u64,
    /// 前一行哈希（hex；创世为 64 个 `0`）。
    pub prev: String,
    /// 本行哈希 = sha256(prev + raw)（hex）。
    pub hash: String,
    /// 写入时的精确条目 JSON 字节序列（验证不依赖 JSON 规范化）。
    pub raw: String,
}

/// 一帧画面。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ReplayFrame {
    /// 采集时间（unix 毫秒）。
    pub ts_ms: u64,
    /// 图像格式（当前恒 `png`）。
    pub format: String,
    /// base64 图像数据。
    pub data_base64: String,
}
